main.go

```diff
@@ -324,10 +338,68 @@ func main() {
 		os.Exit(1)
 	}
 
+	// Migrate Che related resources labels in order to comply with the custom cache function
+	deploy.MigrateCheResourcesLabels(nonCachingClient)
+
 	// Start the Cmd
 	setupLog.Info("starting manager")
 	if err := mgr.Start(sigHandler); err != nil {
 		setupLog.Error(err, "problem running manager")
 		os.Exit(1)
 	}
 }
+
+func getCacheFunc() (cache.NewCacheFunc, error) {
+	cheInstanceRequirement, err := labels.NewRequirement(deploy.KubernetesInstanceLabelKey, selection.Equals, []string{os.Getenv("CHE_FLAVOR")})
+	if err != nil {
+		return nil, err
+	}
+	cheObjectSelector := labels.NewSelector().Add(*cheInstanceRequirement)
+
+	logrus.Infof("Limit cache by selector: %s", cheObjectSelector.String())
+
+	routeKey := &routev1.Route{}
+	selectors := cache.SelectorsByObject{
+		&appsv1.Deployment{}: {
+			Label: cheObjectSelector,
+		},
+		&corev1.Pod{}: {
+			Label: cheObjectSelector,
+		},
+		&batchv1.Job{}: {
+			Label: cheObjectSelector,
+		},
+		&corev1.Service{}: {
+			Label: cheObjectSelector,
+		},
+		&networkingv1.Ingress{}: {
+			Label: cheObjectSelector,
+		},
+		routeKey: {
+			Label: cheObjectSelector,
+		},
+		&corev1.ConfigMap{}: {
+			Label: cheObjectSelector,
+		},
+		&corev1.Secret{}: {
+			Label: cheObjectSelector,
+		},
+		&corev1.ServiceAccount{}: {
+			Label: cheObjectSelector,
+		},
+		&rbacv1.Role{}: {
+			Label: cheObjectSelector,
+		},
+		&rbacv1.RoleBinding{}: {
+			Label: cheObjectSelector,
+		},
+	}
+
+	if !util.IsOpenShift {
```

Comment:
Let's do in opposite way. Add route if it is openshift

Reply:
This is how we did it in [DWO](https://github.com/devfile/devworkspace-operator/pull/652/commits/43aae0090116d23527c4a1faa1cb32a685d0e3ed#diff-964e351ee2375d359c78d69e514c4edc42577219761c4475f391ed2daf715e51R76-R88), and I think the approach here is better -- the struct for constructing this selector is half-internal and so cannot be easily instantiated.

I opened https://github.com/kubernetes-sigs/controller-runtime/issues/1697 but the fix is not present in the version of controller-runtime used here.